File: server/controllers/pregame_controller.rb
class PregameController < AppController

  def initialize(app)
    super(app)
    @players_collection = PlayersCollection.get_instance
  end

  def name
    :pregame
  end

  def closed_conn(conn)
    p = @app.get_client(conn)
    
    unless p.nil? or p.room.nil?
      p.room.remove_player(p)
      if p.room.players.count { |p| !p.is_a?(Ai) } < 1 #remove sala se nao houver mais jogadores conectados nela
        RoomsCollection.get_instance.rem(p.room)
        @app.controllers[:rooms].update_list #envia lista atualizada para todos os jogadores conectados
      else
        update_list(p.room)
      end
      
    end 
   
  end

  def change_color(conn,msg)
    p = @app.get_client(conn)
    colors = get_players_color(p.room,p)
    
    p.color = msg['color'].to_i unless colors.include?(msg['color'].to_i)
    update_player(p)
  end

  def toggle_state(conn,msg)
    p = @app.get_client(conn)
    p.ready = !p.ready
    
    update_player(p)
    
    #verifica se existem pelo menos dois jogadores na sala e se estao todos prontos, e inicia partida
    room = p.room
    if room.players.count > 3 and room.all_ready?
      update_list(room)
      @app.send(room.players,Message.new('pregame','close'))
      puts "Iniciando partida na sala #{room.to_s}..."
      puts "Para os jogadores #{room.players.inspect}"
      @app.controllers[:game].start_game(room)
    end
        
  end
  
  def add_ia(conn,msg)
    p = @app.get_client(conn)
    room = p.room
    adiciona_bot(room) if p.room.owner == p
    update_list(room) if p.room.owner == p
  end
  
  def adiciona_bot(room)
    return nil if room.players.count == 6

    bot = Ai.new(@app.controllers[:game],'BOT')
    bot.room = room
    bot.ready = true
    set_color(bot)
    @app.bind_client(bot.conn,bot)
    @players_collection.add(bot)
    room.players << bot
  end

  def show(player,room)  
    set_color(player)
    update_list(room)
  
    #envia mensagem para abrir modal Pregame ao cliente que juntou-se a sala
    @app.send(player,Message.new('pregame','open'))
    @app.send(player,Message.new('pregame','set_owner',{'is_owner' => true})) if room.players.size == 1
    @app.send(player,Message.new('pregame','set_owner',{'is_owner' => false})) unless room.players.size == 1
  end
  
  private
  def update_player(p)
    @app.send(p.room.players,Message.new('pregame','update_player',{
      'index' => p.room.get_index(p),
      'nick' => p.nick,
      'color' => p.color,
      'ready' => p.ready
    }))
  end
  
  def update_list(room)
    #envia lista de jogadores atualizadas para todos os integrantes da sala    
    players = []
      
    room.players.each do |p|
      players.push({
          'nick' => p.nick,
          'color' => p.color,
          'ready' => p.ready
        }
      )
    end
    
    params = {
      'id' => room.id,
      'name' => room.name,
      'owner' => room.owner.id,
      'size' => '6',
      'players' => players
      }
    @app.send(room.players,Message.new('pregame','update',params))
  end
  
  def get_players_color(room,player=nil)
    unless player.nil?
      players = room.players.clone
      players.delete(player)
    end
    players.map(&:color)
  end
  
  def set_color(player)
    players_color = get_players_color(player.room,player)
    total_colors = Definitions.get_instance.colors['colors'].count
    i = 0
    while players_color.include?(player.color)
      player.color = (i%total_colors)+1
      i += 1
    end
  end
  
end

File: client/js/controllers/pregamecontroller.js
/* controlador da janela pré-jogo, onde se definem os jogadores e as cores */
  var ChangeColorMessage, PregameController, ToggleStateMessage;

  ChangeColorMessage = (function() {

    function ChangeColorMessage(i) {
      this.controller = 'pregame';
      this.action = 'change_color';
      this.params = {
        'color': i
      };
    }

    return ChangeColorMessage;

  })();

  ToggleStateMessage = (function() {

    function ToggleStateMessage() {
      this.controller = 'pregame';
      this.action = 'toggle_state';
      this.params = {};
    }

    return ToggleStateMessage;

  })();
  
  AddIaMessage = (function() {

    function AddIaMessage() {
      this.controller = 'pregame';
      this.action = 'add_ia';
      this.params = {};
    }

    return AddIaMessage;

  })();

  PregameController = (function() {

    function PregameController(app) {
      var _this = this;
      this.app = app;
      this.controllerName = 'pregame';
      this.modal = $('.modal-div#pregame');
      this.players = this.modal.find('ul#players');
      this.colorSelect = this.modal.find('select[name=colors]');
      this.readyButton = this.modal.find('button#ready');
      this.addIaButton = this.modal.find("button#addia");
      this.title = null;
      this.list = null;
      this.is_owner = false;
      this.colorSelect.change(function(eventObject) {
        var index;
        index = $(eventObject.target).val();
        return _this.app.conn.send(new ChangeColorMessage(index));
      });
      this.readyButton.click(function(eventObject) {
        return _this.app.conn.send(new ToggleStateMessage());
      });
      
    }

    PregameController.prototype.update_ia_button = function() {
      var _this = this;
      
      if(!this.is_owner) {
        this.addIaButton.hide();
      }else {
        this.addIaButton.click(function() {
          _this.app.conn.send(new AddIaMessage());
        });
      }
    };

    PregameController.prototype.open = function(msg) {    
      var color, i, _ref;
      this.colorSelect.html('');
      _ref = this.app.controllers['definitions'].get_colors();
      for (i in _ref) {
        color = _ref[i];
        this.colorSelect.append("<option value=" + i + " style=\"color: " + color.hex + "\">" + color.name + "</option>");
      }
      return this.app.openModal(this.modal);
    };

    PregameController.prototype.close = function(msg) {
      return this.app.closeModal(this.modal);
    };

    PregameController.prototype.update = function(msg) {
      this.modal.find('div#modal-title').html(msg.name);
      this.players.html('');
      this.list = msg.players;
      return this.update_list();
    };

    PregameController.prototype.set_owner = function(msg) {
      console.log(msg.is_owner);
      this.is_owner = msg.is_owner;
      this.update_ia_button();
    };

    PregameController.prototype.update_player = function(msg) {
      this.list[msg.index].nick = msg.nick;
      this.list[msg.index].color = msg.color;
      this.list[msg.index].ready = msg.ready;
      return this.update_list();
    };

    PregameController.prototype.update_list = function(list) {
      var colors, hex, p, _i, _len, _ref, _results, line;
      this.players.html('');
      _ref = this.list;
      _results = [];
      for (_i = 0, _len = _ref.length; _i < _len; _i++) {
        p = _ref[_i];
        colors = this.app.controllers['definitions'].get_colors();
        hex = colors[p.color].hex;
        line = '<li>';
        line += '<div class=\"ready-state\" id="' + p.ready + '">';
        line += p.ready ? 'Pronto' : 'Aguardando';
        line += "</div><div class=\"color\" style=\"background-color: " + hex + "\">&nbsp;</div>";
        line += "</div> " + p.nick + "</li>"
        this.players.append(line);
      }
    };

    return PregameController;

  })();

